src/app/util/ember-compatibility-functions.h

```diff
@@ -0,0 +1,115 @@
+
+/**
+ *  Fetch attribute value and version info and write to the AttributeReport provided.
+ *  The ReadSingleClusterData will do everything required for encoding an attribute, i.e. it will try to put one or more
+ * AttributeReportIB to the AttributeReportIBs::Builder.
+ *  When the endpoint / cluster / attribute data specified by aPath does not exist, corresponding interaction
+ * model error code will be put into aAttributeReports, and CHIP_NO_ERROR will be returned. If the data exists on the server, the
+ * data (with tag kData) and the data version (with tag kDataVersion) will be put into aAttributeReports. TLVWriter error will be
+ * returned if any error occurred while encoding these values. This function is implemented by CHIP as a part of cluster data
+ * storage & management.
+ *
+ *  @param[in]    aSubjectDescriptor    The subject descriptor for the read.
+ *  @param[in]    aPath                 The concrete path of the data being read.
+ *  @param[in]    aAttributeReports      The TLV Builder for Cluter attribute builder.
+ *
+ *  @retval  CHIP_NO_ERROR on success
+ */
+CHIP_ERROR ReadSingleClusterData(const Access::SubjectDescriptor & aSubjectDescriptor, bool aIsFabricFiltered,
+                                 const ConcreteReadAttributePath & aPath, AttributeReportIBs::Builder & aAttributeReports,
+                                 AttributeValueEncoder::AttributeEncodeState * apEncoderState);
+
+/**
+ * Returns the metadata of the attribute for the given path.
+ *
+ * @retval The metadata of the attribute, will return null if the given attribute does not exists.
+ */
+const EmberAfAttributeMetadata * GetAttributeMetadata(const ConcreteAttributePath & aPath);
+
+/**
+ * TODO: Document.
+ */
+CHIP_ERROR WriteSingleClusterData(const Access::SubjectDescriptor & aSubjectDescriptor, const ConcreteDataAttributePath & aPath,
+                                  TLV::TLVReader & aReader, WriteHandler * apWriteHandler);
+
+/**
+ * Check if the given cluster has the given DataVersion.
+ */
+bool IsClusterDataVersionEqual(const ConcreteClusterPath & aConcreteClusterPath, DataVersion aRequiredVersion);
+
+/**
+ * Returns true if device type is on endpoint, false otherwise.
+ */
+bool IsDeviceTypeOnEndpoint(DeviceTypeId deviceType, EndpointId endpoint);
+
+/**
+ * Returns the event support status for the given event, as an interaction model status.
+ */
+Protocols::InteractionModel::Status CheckEventSupportStatus(const ConcreteEventPath & aPath);
+
+} // namespace app
+} // namespace chip
+
+/** @brief Reporting Attribute Change
+ *
+ * This function is called by the framework when an attribute managed by the
+ * framework changes.  The application should call this function when an
+ * externally-managed attribute changes.  This function triggers attribute
+ * reports for subscriptions as needed.
+ */
+void MatterReportingAttributeChangeCallback(chip::EndpointId endpoint, chip::ClusterId clusterId, chip::AttributeId attributeId);
```

Comment:
these seem completely unrelated (and their definition was in reporting/reporting.h).

Adding this comment as I am considering splitting this to have a bit less circular dependency in things.

Reply:
Created 'reporting.cpp' instead now, to have things isolated.

The odd dependnecy is only because ember data model version is stored in generated code, so the header and cpp files cannot reside together now (global code needs to call 'notify' and actual implementation depends on generated code).